rancher/rancher
# feat: Allows configuration of the Service

## Issue: 
https://github.com/rancher/rancher/issues/16061
 
## Problem
When deploying to GKE using the chart provided service, the load balancer fails to deploy. I specifically noticed this when running the setup below.

* Private GKE cluster
* Using the GKE (Google) Ingress  (Not ``nginx`` or ``trafeik``.)
* Binding the ingress with a static IP address and DNS record already provisioned by terraform using extra annotations (https://cloud.google.com/kubernetes-engine/docs/concepts/ingress-xlb#static_ip_addresses_for_https_load_balancers)
* Using Google Managed TLS Certificates using extra annotations (https://cloud.google.com/kubernetes-engine/docs/how-to/managed-certs#setting_up_a_google-managed_certificate)
* Terminating SSL at load balancer

Basically it seems that Google Managed certificates / GKE ingress can only run with NodePort services. Currently the chart defaults to ``ClusterIP`` and doesn't give you any manner of overriding this.

We also had problems applying IAP (Identity Aware Proxy) and service timeouts (google kills connections after 30 seconds by default and this breaks websockets) to our Rancher service. This is because the current rancher ``service.yaml`` file does not allow for customisation of service annotations.
 
## Solution
This PR contains a fix to allow the ``type`` attribute on the service to be provided by the user and to utilise that if it is passed. If it is not passed, the current behaviour will apply and a ``ClusterIP`` will be used.

It also allows a user to pass a block of service annotations to apply to the service. If no service annotations are passed, then the current behaviour will apply an no annotations will be provided for the service.
 
## Testing


## Engineering Testing
### Manual Testing
* Ran the changes above in a local check out of the chart and deployed successfully using a ``NodePort`` service type.

### Automated Testing
This PR has added helm tests for the feature added.

### Regressions Considerations
No regressions should be introduced here since the current behaviours will persist unless specifically overridden.
